### PR TITLE
Add liveness probe for VPA components

### PR DESCRIPTION
**Pull request by andrerun**
**How to categorize this PR?**

/area auto-scaling
/kind enhancement

**What this PR does / why we need it**:
Adds liveness probe for VPA components. In the past, we have seen multiple situations where a VPA component got stuck, not providing target workloads with recommendations/updates. This change leverages the existing k8s liveness infrastructure to remediate such situations. 

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Release note**:

```other operator
VPA components do now have a liveness probe defined.
```

### CODE REVIEW

File: pkg/operation/botanist/component/vpa/admissioncontroller.go [hunk 1]
```diff
@@ -208,6 +208,7 @@ func (v *vpa) reconcileAdmissionControllerDeployment(deployment *appsv1.Deployme
 						fmt.Sprintf("--port=%d", admissionControllerPort),
 						"--register-webhook=false",
 					},
+					LivenessProbe: getDefaultLivenessProbe(),
```
Comment (voelzmo) [inline pkg/operation/botanist/component/vpa/admissioncontroller.go]: I'm not sure we can add this to the AdmissionController with the standard settings safely. While the Recommender and the Update have a 60 second loop in which they update their healthiness status, the AdmissionController does this [whenever it serves a request](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/admission-controller/main.go#L124), where we don't really have any guarantees, right?

File: pkg/operation/botanist/component/vpa/general.go [hunk 2]
```diff
@@ -27,6 +30,31 @@ import (
...
+				Path:   "/health-check",
+				Port:   intstr.FromString(metricsPortName),
+				Scheme: corev1.URISchemeHTTP,
+			},
+		},
+		// Typically, the short-term impact of losing VPA is low
+		// So, we can afford relaxed liveness timing and avoid unnecessary restarts aggravating high load situations
+		InitialDelaySeconds: 120,
+		PeriodSeconds:       60,
+		TimeoutSeconds:      30,
```
Comment (voelzmo) [inline pkg/operation/botanist/component/vpa/general.go:49]: Do we really think that this request can/should take that long and we still consider the component alive? Even 5 seconds sounds like a long time to me.

File: pkg/operation/botanist/component/vpa/general.go [hunk 3]
```diff
@@ -27,6 +30,31 @@ import (
 	"k8s.io/utils/pointer"
 )
 
+const (
+	metricsPortName = "metrics"
+)
+
+// getDefaultLivenessProbe returns a new instance of the default liveness probe configuration, used for VPA components
```
Comment (voelzmo) [inline pkg/operation/botanist/component/vpa/general.go]: I don't think we need to comment private functions, let's remove this comment.

File: pkg/operation/botanist/component/vpa/general.go [hunk 4]
```diff
@@ -27,6 +30,31 @@ import (
...
+func getDefaultLivenessProbe() *corev1.Probe {
+	// Return a copy as insurance against unintentional aliasing
+	return &corev1.Probe{
+		ProbeHandler: corev1.ProbeHandler{
+			HTTPGet: &corev1.HTTPGetAction{
+				Path:   "/health-check",
+				Port:   intstr.FromString(metricsPortName),
```
Comment (voelzmo) [inline pkg/operation/botanist/component/vpa/general.go:41]: Nice usage of named ports here! 👍

File: pkg/operation/botanist/component/vpa/general.go [hunk 5]
```diff
@@ -27,6 +30,31 @@ import (
 	"k8s.io/utils/pointer"
 )
 
+const (
+	metricsPortName = "metrics"
+)
+
+// getDefaultLivenessProbe returns a new instance of the default liveness probe configuration, used for VPA components
+func getDefaultLivenessProbe() *corev1.Probe {
+	// Return a copy as insurance against unintentional aliasing
```
Comment (voelzmo) [inline pkg/operation/botanist/component/vpa/general.go]: I'm not sure what that comment means – what is it that we're avoiding here?

**Review by voelzmo — CHANGES_REQUESTED** [review]

File: pkg/operation/botanist/component/vpa/admissioncontroller.go (hunk 1 above)
Comment (andrerun) [reply pkg/operation/botanist/component/vpa/admissioncontroller.go]: After looking more closely at the admission controller, I see that its liveness probe does two things:
1) Do some work to determine if the period since last activity exceed the specified timeout period (one minute)
2) Throw that information away, and return HTTP 200 "OK" unconditionally.

I don't see how the liveness probe would have any effect, except when the probe itself times out. That could happen, but, networking issues aside, I expect it to be quite rare (e.g. extreme CPU starvation/exclusion due to some high priority component spinning in multiple loops, or kernel/driver issue resulting in a tight loop/deadlock; also, a hypothetical extremely slow memory allocation, due to GC or system heap compaction - don't know if that's a thing on Linux). Basically the thread scheduler on the host being prevented from effectively scheduling threads. I've seen many such issues on my desktop machines over the years, but I think they should be rare on server class machines. I'd say that at this point in time, we're implementing admission controller probing as much to benefit from potential future enhancements to the probe, as from its current function.

File: pkg/operation/botanist/component/vpa/general.go (hunk 2 above)
Comment (andrerun) [reply pkg/operation/botanist/component/vpa/general.go:49]: I chose those values, accounting for how VPA outage doesn't cause much short term problems. My thinking was, a one second probe makes a difference over a 30 second one, only when the VPA controller is being overloaded with work. In those circumstances, a strictly timed probe is likely to become part of the problem by introducing costly restarts to a system which is already struggling.

A schedulable thread not getting scheduled within 1 second (or 3) is a rare event, in terms of percentage of scheduling time slices. But it's not that rare in human time. Pauses due to memory compaction, intermittent network faults, CPU starvation due to higher priority threads, intermittent hardware faults, vmotion completion, all easily exceed 1 second. As a general rule, I try to avoid load amplification when I can, in fear of cascading failures. It's strictly a defensive practice, rarely driven by case-specific concerns.

File: pkg/operation/botanist/component/vpa/general.go (hunk 5 above)
Comment (andrerun) [reply pkg/operation/botanist/component/vpa/general.go]: I'm avoiding what would happen with the alternative of reusing the same static object (vs. copy), and delegating the responsibility for not changing it to the caller:
a = getDefaultLivenessProbe()
b = getDefaultLivenessProbe()
a.ProbeHandler.Port = 80 // b.ProbeHandler.Port is unintentionally changed to 80

Basically paying compute tax to work around golang's lack of compound object "const-ness".

File: pkg/operation/botanist/component/vpa/general.go (hunk 3 above)
Comment (andrerun) [reply pkg/operation/botanist/component/vpa/general.go]: Done.

File: pkg/operation/botanist/component/vpa/general.go [hunk 6]
```diff
@@ -17,6 +17,9 @@ package vpa
 import (
 	"fmt"
 
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
```
Comment (ialidzhikov) [inline pkg/operation/botanist/component/vpa/general.go]: Nit: You could move to the 3rd import group.

File: pkg/operation/botanist/component/vpa/general.go (hunk 5 above)
Comment (ialidzhikov) [reply pkg/operation/botanist/component/vpa/general.go]: I cannot follow what you mean. The case you described above does not seem to be right:
```golang
func main() {
	a := newDefaultLivenessProbe()
	b := newDefaultLivenessProbe()

	a.ProbeHandler.HTTPGet.Port = intstr.FromInt(80)
	fmt.Printf("%+v\n", a.ProbeHandler.HTTPGet.Port)
	fmt.Printf("%+v\n", b.ProbeHandler.HTTPGet.Port)
}
```

This outputs

```
{Type:0 IntVal:80 StrVal:}
{Type:1 IntVal:0 StrVal:metrics}
```

=> `b.ProbeHandler.HTTPGet.Port` does not get unintentionally changed

---

I guess you could remove the confusing comment.
```suggestion
```

File: pkg/operation/botanist/component/vpa/updater.go [hunk 7]
```diff
@@ -23,7 +23,6 @@ import (
 	"github.com/gardener/gardener/pkg/operation/botanist/component"
 	"github.com/gardener/gardener/pkg/utils"
 	gutil "github.com/gardener/gardener/pkg/utils/gardener"
-
```
Comment (ialidzhikov) [inline pkg/operation/botanist/component/vpa/updater.go]: Nit: You could revert this change.

File: pkg/operation/botanist/component/vpa/admissioncontroller.go [hunk 8]
```diff
@@ -208,6 +208,7 @@ func (v *vpa) reconcileAdmissionControllerDeployment(deployment *appsv1.Deployme
 						fmt.Sprintf("--port=%d", admissionControllerPort),
 						"--register-webhook=false",
 					},
+					LivenessProbe: newDefaultLivenessProbe(),
```
Comment (ialidzhikov) [inline pkg/operation/botanist/component/vpa/admissioncontroller.go:211]: Don't we need a readiness probe as well for the VPA admission controller as it is a component that also receives traffic?

Comment (voelzmo) [reply pkg/operation/botanist/component/vpa/admissioncontroller.go:211]: You're right, the admission controller doesn't have a readiness probe yet. As we had seen at least one occurrence where a VPA component was stuck, we wanted to integrate the already existing health check function as liveness to resolve this. If we have a good candidate for readiness we can also add this to the admission controller – maybe in an additional PR?

File: pkg/operation/botanist/component/vpa/general.go (hunk 5 above)
Comment (voelzmo) [reply pkg/operation/botanist/component/vpa/general.go]: @andrerun the scenario you're describing would happen when the `newDefaultLivenessProbe()` implementation would look like this – in all other cases this should work as @ialidzhikov's example run showed.
```
var p = corev1.Probe{
	ProbeHandler: corev1.ProbeHandler{
		HTTPGet: &corev1.HTTPGetAction{
			Path:   "/health-check",
			Port:   intstr.FromString("metrics"),
			Scheme: corev1.URISchemeHTTP,
		},
	},
	InitialDelaySeconds: 120,
	PeriodSeconds:       60,
	TimeoutSeconds:      30,
	SuccessThreshold:    1,
	FailureThreshold:    3,
}

func newDefaultLivenessProbe() *corev1.Probe {
	return &p
}
```
So I guess let's remove the comment and keep the implementation, this is fine like it is.

File: pkg/operation/botanist/component/vpa/admissioncontroller.go (hunk 1 above)
Comment (voelzmo) [reply pkg/operation/botanist/component/vpa/admissioncontroller.go]: Ah, you're right, [the `false` parameter passed into the healthcheck](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/admission-controller/main.go#L78) leads to [ignoring if the timeout happened or not](https://github.com/kubernetes/autoscaler/blob/ef2d9e7169295724ce9e87d21235698aa9d2b60c/vertical-pod-autoscaler/pkg/utils/metrics/healthcheck.go#L55). So that means my above comment about a risk of timeout is not correct. We can certainly add this to the admission-controller, then, even if it will never intentionally return an unhealthy state.

File: pkg/operation/botanist/component/vpa/admissioncontroller.go (hunk 8 above)
Comment (ialidzhikov) [reply pkg/operation/botanist/component/vpa/admissioncontroller.go:211]: Ok, let's postpone it to potentially other PR. I now read that it is not recommended to set the same configuration for both the probes - the liveness probe is much more "aggressive" -> it forces a container restarts and shadows the effect of a failing readiness probe.

**Comment by rfranzke** [comment]
What is the status of this PR? @andrerun do you plan to address the comments?

**Comment by voelzmo** [comment]
@rfranzke Andrey was on vacation and is back today.

File: pkg/operation/botanist/component/vpa/general.go (hunk 5 above)
Comment (andrerun) [reply pkg/operation/botanist/component/vpa/general.go]: @voelzmo @ialidzhikov My goal with the comment was to highlight the intent of returning a fresh copy of the object upon each call, as opposed to reusing a singleton static instance. I felt I needed the comment, especially because I remembered seeing the other approach also used in Gardener source code in the past - large mutable objects stored in static variables. However, after prompted by your remarks, I did look more closely, and it turns out that the pattern is only used in tests, where [aliasing](https://en.wikipedia.org/wiki/Aliasing_(computing)) was not a risk.

Generally I prefer to err on the side of more code comments, to reduce the risk that I fall prey to the [curse of knowledge bias](https://en.wikipedia.org/wiki/Curse_of_knowledge). But, seeing that you both agree that the intent of that particular piece of code is immediately apparent, off with the comment.

File: pkg/operation/botanist/component/vpa/updater.go (hunk 7 above)
Comment (andrerun) [reply pkg/operation/botanist/component/vpa/updater.go]: @ialidzhikov The change was made by some of the tools in GoLand's automatic formatting pipeline. Do we have an established convention about how we organise imports? I'd like to configure my tooling accordingly, otherwise I'll be running into the same issue ad infinitum.

File: pkg/operation/botanist/component/vpa/general.go (hunk 6 above)
Comment (andrerun) [reply pkg/operation/botanist/component/vpa/general.go]: @ialidzhikov Can we shake hands on moving the 3rd group (third-party libraries) so it becomes a 2nd group?

The insertion was made automatically by GoLand, per default configuration. I don't care how imports are organised, as long as they are sorted lexicographically. However, if I absolutely had to take a side,  I'd side with GoLand's choice: Most highly reusable imports first (standard language library). Then moderately reusable imports (domain-specific libraries). Then least reusable code - application's own peer modules.

If the team has already settled upon the other ordering as convention (stdlib -> app peer imports -> 3rd party libs), please let me know so I can configure my tooling accordingly and not fight it in the future.

**Review by ialidzhikov — COMMENTED** [review]
/approve

@voelzmo can you recheck this PR? Thanks in advance!

**Review by voelzmo — APPROVED** [review]
/lgtm thanks!
